README.md
# Presentation Plugin

The **Presentation** Plugin is an extension for [Grav CMS](http://github.com/getgrav/grav), and provides a simple way of creating fullscreen slideshows that can be navigated two-dimensionally, using the [Reveal.js](https://github.com/hakimel/reveal.js/)-library.

At its core the plugin facilitates efficient handling of content for use with the library. You can utilize Reveal.js however you want through custom initialization, and still leverage the plugin's content-handling. A [demo is available](https://olevik.me/staging/grav-skeleton-presentation/) and a [Skeleton](https://github.com/OleVik/grav-skeleton-presentation/releases), as well as the [demo content](https://github.com/OleVik/grav-skeleton-presentation/tree/master/pages).

## Features

- Presentations through two-dimensional slideshows
- Responsive, multi-device capable styling
- Granular control over presentation and slides
  - Cascading-styles and options with Shortcodes and the Admin-plugin
- Portable content: Everything is contained in Markdown, including settings
- Flexible, ambigious, recursive Page-structure
- Extendable through your own theme or plugin
- Print-friendly presentations, with or without notes or in text-only mode
- Presenter-mode: A modern, powerful, easy-to-use alternative to PowerPoint
  - Include notes with your presentation
  - Synchronize Presenter-mode and the Presentation locally or remotely

## Installation

Installing the presentation-plugin can be done in one of two ways. The GPM (Grav Package Manager) installation method enables you to quickly and easily install the plugin with a simple terminal command, while the manual method enables you to do so via a zip file.

### GPM Installation (Preferred)

The simplest way to install this plugin is via the [Grav Package Manager (GPM)](http://learn.getgrav.org/advanced/grav-gpm) through your system's terminal (also called the command line). From the root of your Grav install type:

    bin/gpm install presentation

This will install the Presentation-plugin into your `/user/plugins` directory within Grav. Its files can be found under `/your/site/grav/user/plugins/presentation`.

### Manual Installation

To install this plugin, just download the zip version of this repository and unzip it under `/your/site/grav/user/plugins`. Then rename the folder to `presentation`. You can find these files on [GitHub](https://github.com/ole-vik/grav-plugin-presentation) or via [GetGrav.org](http://getgrav.org/downloads/plugins#extras).

You should now have all the plugin files under

    /your/site/grav/user/plugins/presentation

> NOTE: This plugin is a modular component for Grav which requires [Grav](http://github.com/getgrav/grav) and the [Error](https://github.com/getgrav/grav-plugin-error) and [Problems](https://github.com/getgrav/grav-plugin-problems) to operate.

### Requirements

**This plugin will only work with [Grav v1.6](https://github.com/getgrav/grav/tree/1.6) or higher, as it requires [PHP v7.1.3](http://php.net/manual/en/migration71.new-features.php) or higher.**

## Configuration

Before configuring this plugin, you should copy the `user/plugins/presentation/presentation.yaml` to `user/config/plugins/presentation.yaml` and only edit that copy.

Here is the default configuration and an explanation of available options:

```yaml
# Enable Plugin if true, disable if false
enabled: true
# Theme to use from Reveal.js (https://github.com/hakimel/reveal.js/#theming)
theme: moon
# Order of how pages are rendered
order:
  by: folder
  dir: asc
# Load all registered CSS and JS assets
all_assets: false
# Include Theme's custom.css
theme_css: true
# Enable Reveal.js CSS
builtin_css: true
# Enable Reveal.js JS
builtin_js: true
# Enable Plugin's CSS
plugin_css: true
# Enable Plugin's JS
plugin_js: true
# Enable Plugin's dynamic text sizing
textsizing: true
# Synchronize Slide-navigation
sync: "none"
# URL Route to use for Poll-sync
api_route: "presentationapi"
# Poll-sync timeout in milliseconds
poll_timeout: 2000
# Poll-sync retry limit
poll_retry_limit: 10
# Enable Poll-sync token-authorization
token_auth: false
# Poll-sync token to use for authorization
token: Hd4HFdPvbpKzTqz
# Enable Save Content button and CTRL+SHIFT+S combo
admin_async_save: false
# Enable Save Content when typing
admin_async_save_typing: false
# Twig-template to inject below content
footer: ""
# Enable onLoad transition
transition: true
# Enable Plugin's shortcodes
shortcodes: true
# Default class for Presentation-shortcode
shortcode_classes: "presentation-iframe"
# Unwrap images from paragraph
unwrap_images: true
# Class to use for Content building
content: "Content"
# Class to use for Content parsing
parser: "Parser"
# Class to use for Styles, Classes, and Data management
transport: "Transport"
# Class to use for Shortcode parsing
shortcode_parser: RegularParser
# Breakpoints for responsive textsizing
breakpoints:
  240: "16"
  320: "20"
  576: "24"
  768: "28"
  992: "32"
  1200: "36"
  1600: "40"
# Styles to use as defaults for Presentations
style:
  ...
# Dynamic text scaling to use as defaults for Presentations
textsize:
  scale: string
  modifier: float
# Stack slides horizontally or vertically (default), on thematic breaks
horizontal: false
# Options to pass to Reveal.js
options:
  width: "100%"
  height: "100%"
  margin: "0"
  minScale: "1"
  maxScale: "1"
  transition: "fade"
  controlsTutorial: "false"
  history: "true"
  display: "flex"
  pdfSeparateFragments: false
```

All configuration-options available to the Reveal.js-library can be configured through `options`, see its [documentation for available options](https://github.com/hakimel/reveal.js#configuration).

### Page-specific configuration

Any configuration set in `presentation.yaml` can be overridden through a Page's FrontMatter, like this:

```yaml
---
title: Alice’s Adventures in Wonderland
presentation:
  order:
    by: date
    dir: desc
  options:
    transition: "fade"
---

```

## Usage

The Page-structure used in the Presentation-plugin is essentially the same as normally in Grav, with a few notable exceptions: Any horizontal rule, `---` in Markdown and `<hr />` in HTML, is treated as a _thematic break_, as it is defined in HTML5. This means that every Page in Grav is treated as a normal, _horizontal Slide_ when the Plugin iterates over them, but a thematic break creates a _vertical Slide_.

You can have as many Pages below the root-page as you want, each of them will be treated as a Slide. When you create thematic breaks within the Page, the Slides are then created _below_ the Page itself -- accommodating Reveal.js' two-dimensional slideshows.

For using Reveal.js itself, see their documentation on [Getting Started with Presenting](http://help.slides.com/knowledgebase/articles/334920) and [Speaker View](http://help.slides.com/knowledgebase/articles/333923-speaker-view).

### Nomenclature

With Reveal.js the presentation is not entirely linear. Rather, it has a linear, left-to-right set of sections that each make up a slide, and can have additional slides going downwards. Thus you can progress through the presentation linearly starting at each section, moving downwards until the end, and continuing onto the next section, or move between them as you choose.

Further, there are [Fragments](https://github.com/hakimel/reveal.js#fragments) that can be used within each slide. These reveal linearly like slides, but make one element appear at a time rather than the full contents of the slide.

### Structure

```
/user/pages/book
├── presentation.md
├── 01.down-the-rabbit-hole
│   └── slide.md
├── 02.advice-from-a-caterpillar
│   └── slide.md
├── 03.were-all-mad-here
│   └── slide.md
├── 04.a-mad-tea-party
│   └── slide.md
├── 05.the-queens-crocquet-ground
│   └── slide.md
├── 06.postscript
└───└── slide.md
```

As seen in this example structure, only the initial page uses the `presentation.html.twig`-template. The template used for child-pages is `slide.html.twig`, though the content of these pages are processed regardless of template. Naming them `slide.md` enables the blueprints for slides in the Admin-plugin. The plugin defines the `presentation.html.twig`-template, but you can override it through your theme.

### Styling

The plugin emulates the logic of Cascading Style Sheets (CSS), in that pages can be assigned a class, style-property, or be hidden using FrontMatter or shortcodes. This is as simple as using `class: custom-slide-class` in FrontMatter or `[class=custom-slide-class]` with a shortcode in the Markdown-content. Styles are applied the same way, where FrontMatter accepts CSS-properties like this:

```
style:
  color: green
```

That is, mapping each property to a value, not as a list. The same could be set for any single slide using `[style-color=green]`, as described below. Styles are given precedence by where they appear, so the plugins looks for them in this order:

1. Plugin-options
2. Page FrontMatter
3. Child page FrontMatter
4. Page Content (Markdown) as shortcodes

The properties are gathered cumulatively, and a property farther down the chain takes precedence over a property further up.

You can of course also style the plugin using your theme's `/css/custom.css`-file, by targeting the `.reveal`-selector which wraps around all of the plugin's content. This behavior can be enabled or disabled with the `theme_css`-setting. All slides have an `id`-attribute set on their sections, which can be utilized by CSS like this:

```css
.reveal #down-the-rabbit-hole-0 {
  background: red;
}
```

#### Fitting text to a slide

The plugin makes available a method of dynamically scaling text within a slide, which is similar yet distinct from what happens in PowerPoint 2016. Rather than do this scaling entirely automatically, which tends to work poorly across devices and resolutions, you set a _scale_ and an optional _modifier_, eg.:

```
[data-textsize-scale=1.125]
[data-textsize-modifier=1.05]
```

If Textsizing is enabled in the plugin's options and on the Page, the relation between block text -- any text not in a header-element -- and header-text (`h1`, `h2`, `h3`, `h4`, `h5`, `h6`) is determined by the `textsize-scale`-property. That is, the size of the header-element's text relative to the base font-size.


CHANGELOG.md
# v4.0.2
## 20-01-2021

1. [](#improved)
   - Theme-select redundancy

# v4.0.1
## 19-07-2020

1. [](#bugfix)
   - Language-string for advanced options

# v4.0.0
## 15-07-2020

1. [](#new)
   - Stable release of 4.0.0

# v4.0.0-beta.2
## 15-07-2020

1. [](#improved)
   - Twig-extensions namespacing

# v4.0.0-beta.1
## 12-07-2020

1. [](#bugfix)
   - Re-processing content with Parsedown broke slides
2. [](#new)
   - `process`-option

# v3.1.4-beta.1
## 12-07-2020

1. [](#bugfix)
   - Flexibility in declaring styles from config and FrontMatter
   - Reinstate `reveal_init` initialization
2. [](#improved)
   - Stricter type-checking on styles

# v3.1.3
## 23-05-2020

1. [](#bugfix)
   - Fix transport for styles

# v3.1.2
## 10-02-2020

1. [](#bugfix)
   - Reduce reliance on config()

# v3.1.1
## 09-02-2020

1. [](#bugfix)
   - Handle null-values saved by Admin

# v3.1.0
## 09-02-2020

1. [](#new)
   - Enable ParsedownExtra when set in system.yaml
2. [](#bugfix)
   - Variable in blueprints.yaml and languages.yaml

# v3.0.2
## 01-10-2019

1. [](#bugfix)
   - Testing-parameter is useless, let me have Semantic Versioning!

# v3.0.1
## 01-10-2019

1. [](#new)
   - Release 3.0.1

# v3.0.0-beta.1
## 15-09-2019

1. [](#improved)
   - Align content processing with Shortcode Core (https://github.com/getgrav/grav-plugin-shortcode-core/issues/75#issuecomment-531596152)

# v2.1.0
## 09-08-2019

1. [](#improved)
   - Defer assets loading (#11)
   - README
   - Blueprints

# v2.0.3
## 17-07-2019

1. [](#bugfix)
   - Blank value fallback in shortcode-processing

# v2.0.2
## 22-06-2019

1. [](#improved)
   - Cache-manipulation
   - Config-retrieval

# v2.0.1
## 10-06-2019

1. [](#bugfix)
   - API-interface alignment

# v2.0.0
## 10-06-2019

1. [](#new)
   - Backward-incompatible API-changes
   - Simplified shortcode-processing
2. [](#improved)
   - Prevent Twig-caching
   - Fragment-parsing
   - Shortcode-handling through Transport
   - Styles-handling through Parser
   - API-selection in blueprints
   - Default textsizing-scale and breakpoint-check
3. [](#bugfix)
   - Data-shortcode handling

# v1.7.4
## 03-06-2019

1. [](#new)
   - Trim quotes in iframe-shortcode

# v1.7.3
## 03-06-2019

1. [](#new)
   - Test GPM wickedness, take 2

# v1.7.2
## 03-06-2019

1. [](#new)
   - Test GPM wickedness

# v1.7.1
## 01-06-2019

1. [](#new)
   - Stable release of refactor

# v1.7.1-beta.10
## 31-05-2019

1. [](#improved)
   - Revert 'Initialize API earlier', do it indirectly

# v1.7.1-beta.9
## 31-05-2019

1. [](#improved)
   - Versions
   - Deprecate 'v' from tags

# v1.7.1-beta.8
## 31-05-2019

1. [](#improved)
   - Initialize API earlier
   - Refactor Parser->interpretShortcodes()
   - Parse Shortcodes with Thunderer/Shortcode
2. [](#new)
   - Option for Shortcode Parser

# v1.7.1-beta.7
## 19-05-2019

1. [](#bugfix)
   - Test composer backport of Shortcode Code

# v1.7.1-beta.6
## 15-05-2019

1. [](#bugfix)
   - Page.find for Presentation-shortcode

# v1.7.1-beta.5
## 15-05-2019

1. [](#improved)
   - Pass Page to Presentation-shortcode template

# v1.7.1-beta.4
## 14-05-2019

1. [](#new)
   - README: Creating a shortcode or shortcode-alias
   - README: Twig
   - README: Contributing
2. [](#improved)
   - Split docs into parts
   - Link Overlay method
3. [](#bugfix)
   - Image-scaling
   - Autoloading

# v1.7.1-beta.3
## 13-05-2019

1. [](#bugfix)
   - Maximize image width

# v1.7.1-beta.2
## 13-05-2019

1. [](#bugfix)
   - Shortcode Core backfill for Packagist

# v1.7.1-beta.1
## 12-05-2019

1. [](#new)
   - Packagist-release
   - Require Shortcode Core as a dependency
   - Link Overlay shortcode
   - Presentation shortcode: `presentation_base_url` replaced by `base_url`
2. [](#improved)
   - PSR-4 handling through Composer

# v1.7.0-beta.1
## 11-05-2019

1. [](#bugfix)
   - Shortcode-parameters
2. [](#improved)
   - Handle iFrame-interaction
   - README: Backgrounds
   - README: Shortcode-parameters

# v1.6.8
## 11-05-2019

1. [](#bugfix)
   - Admin-detection (#8)
   - Slide-rendering (#9)
2. [](#improved)
   - Controls in slide-blueprint

# v1.6.7
## 05-05-2019

1. [](#improved)
   - Box-sixing, margins for blockquote

# v1.6.6
## 20-04-2019

1. [](#new)
   - Version

# v1.6.5
## 19-04-2019

1. [](#new)
   - Option to load all assets
1. [](#improved)
   - Background-image URL and search

# v1.6.4
## 06-04-2019

1. [](#improved)
   - Plugin default style specificity
2. [](#bugfix)
   - Move plugin default-styles to top-level element
   - Revert ordering-query in CSS
   - Print: CSS, reinitialize background images

# v1.6.3
## 17-03-2019

1. [](#bugfix)
   - Correct ordering-query in CSS

# v1.6.2
## 17-03-2019

1. [](#bugfix)
   - Correct ordering in CSS

# v1.6.1
## 16-03-2019

1. [](#bugfix)
   - Compress CSS
   - Slides ordering with z-index

# v1.6.0
## 13-03-2019

1. [](#new)
   - Alignment-options
2. [](#improved)
   - Allow default styles from plugin options
   - Blueprints structure, language-strings
3. [](#bugfix)
   - Horizontal- and textsizing-option logic
   - Changelog-formatting
   - Remove defaults from blueprints

# v1.5.2
## 23-02-2019

1. [](#improved)

presentation.php
<?php
/**
 * Presentation Plugin
 *
 * PHP version 7
 *
 * @category   Extensions
 * @package    Grav
 * @subpackage Presentation
 * @author     Ole Vik <git@olevik.net>
 * @license    http://www.opensource.org/licenses/mit-license.html MIT License
 * @link       https://github.com/OleVik/grav-plugin-presentation
 */
namespace Grav\Plugin;

use Grav\Common\Grav;
use Grav\Common\Plugin;
use Grav\Common\Utils;
use Grav\Common\Uri;
use Grav\Common\Inflector;
use Grav\Common\Page\Page;
use Grav\Common\Page\Pages;
use Grav\Common\Page\Media;
use Grav\Common\Page\Collection;
use RocketTheme\Toolbox\Event\Event;

use Grav\Plugin\PresentationPlugin\API\Content;
use Grav\Plugin\PresentationPlugin\API\Parser;
use Grav\Plugin\PresentationPlugin\API\Transport;
use Grav\Plugin\PresentationPlugin\API\Poll;
use Grav\Plugin\PresentationPlugin\Utilities;

/**
 * Creates slides using Reveal.js
 *
 * Class PresentationPlugin
 *
 * @category Extensions
 * @package  Grav\Plugin
 * @author   Ole Vik <git@olevik.net>
 * @license  http://www.opensource.org/licenses/mit-license.html MIT License
 * @link     https://github.com/OleVik/grav-plugin-presentation
 */
class PresentationPlugin extends Plugin
{

    /**
     * Protected variables
     *
     * @var bool      $cache     Grav cache setting
     * @var Transport $transport Transport API
     * @var Parser    $parser    Parser API
     * @var Content   $content   Content API
     */
    protected $cache;
    protected $cacheTwig;
    protected $transport;
    protected $parser;
    protected $content;

    /**
     * Register intial event and libraries
     *
     * @return array
     */
    public static function getSubscribedEvents()
    {
        include __DIR__ . '/vendor/autoload.php';
        return [
            'onPluginsInitialized' => ['onPluginsInitialized', 0],
        ];
    }

    /**
     * Initialize the plugin and events
     *
     * @return void
     */
    public function onPluginsInitialized()
    {
        if ($this->config->get('plugins.presentation.enabled') != true) {
            return;
        }
        if ($this->config->get('system.debugger.enabled')) {
            $this->grav['debugger']->startTimer('presentation', 'Presentation');
        }
        if ($this->isAdmin() && $this->config->get('plugins.admin')) {
            $this->enable(
                [
                    'onPagesInitialized' => ['handleAPI', 0],
                    'onGetPageTemplates' => ['onGetPageTemplates', 0],
                    'onTwigSiteVariables' => ['twigBaseUrl', 0],
                    'onAssetsInitialized' => ['onAdminPagesAssetsInitialized', 0]
                ]
            );
        }
        $this->cache = $this->grav['config']->get('system.cache.enabled');
        $this->cacheTwig = $this->grav['config']->get('system.pages.never_cache_twig');
        $this->enable(
            [
                'onShortcodeHandlers' => ['onShortcodeHandlers', 0],
                'onPageInitialized' => ['pagePreCache', 0],
                'onPageContentProcessed' => ['pageIteration', -10],
                'onTwigExtensions' => ['onTwigExtensions', 0],
                'onTwigTemplatePaths' => ['templates', 0],
                'onTwigSiteVariables' => ['twigBaseUrl', 0],
                'onPagesInitialized' => ['handleAPI', 0],
                'onAssetsInitialized' => ['onAssetsInitialized', 0],
                'onShutdown' => ['onShutdown', 0]
            ]
        );
        if ($this->config->get('system.debugger.enabled')) {
            $this->grav['debugger']->stopTimer('presentation');
        }
    }

    /**
     * Declare config from plugin-config
     *
     * @return array Plugin configuration
     */
    public function config()
    {
        if ($this->config->get('plugins.presentation')) {
            return (array) $this->config->get('plugins.presentation');
        }
        return false;
    }

    /**
     * Disable caching for related templates
     *
     * @return void
     */
    public function pagePreCache()
    {
        if ($this->grav['page']->template() == 'presentation' || $this->grav['page']->template() == 'slide') {
            $this->grav['config']->set('system.cache.enabled', false);
            $this->grav['config']->set('system.pages.never_cache_twig', true);
        }
    }


    /**
     * Construct the page
     *
     * @return void
     */
    public function pageIteration()
    {
        $grav = $this->grav;
        $config = $this->config();
        if ($grav['page']->template() == 'presentation' || $grav['page']->template() == 'slide') {
            if (!isset($this->grav['twig']->twig_vars['reveal_init'])) {
                $baseUrl = $this->grav['uri']->rootUrl(true);
                $header = (array) $grav['page']->header();
                if (isset($header['presentation'])) {
                    $config = Utils::arrayMergeRecursiveUnique(
                        $config,
                        $header['presentation']
                    );
                }
                $this->transport = $this->getAPIInstance(
                    $config['transport']
                );
                $this->parser = $this->getAPIInstance(
                    $config['parser'],
                    $config,
                    $this->transport
                );
                $this->content = $this->getAPIInstance(
                    $config['content'],
                    $this->grav,
                    $config,
                    $this->parser,
                    $this->transport
                );
                $styles = $this->config->get('plugins.presentation.style') ??
                    $this->config->get('plugins.presentation.styles') ??
                    [];
                if (!empty($styles) && is_array($styles) && Utils::arrayIsAssociative($styles)) {
                    $this->parser->processor(
                        $styles,
                        'presentation',
                        (array) $grav['page']
                    );
                }
                $tree = $this->content->buildTree($grav['page']->route());
                $slides = $this->content->buildContent($tree);
                $grav['page']->setRawContent($slides);
                $menu = $this->content->buildMenu($tree);
                $menu = Utilities::flattenArray($menu, 1);
                $options = Utilities::parseAmbiguousArrayValues(
                    $this->config->get('plugins.presentation.options')
                );
                $options = json_encode($options, JSON_PRETTY_PRINT);
                $breakpoints = json_encode(
                    $this->config->get('plugins.presentation.breakpoints')
                );
                $this->grav['twig']->twig_vars['reveal_init'] = $options;
                $grav['assets']->addInlineJs('const reveal_init = ' . $options . ';', null, 'presentation');
                $this->grav['twig']->twig_vars['presentation_menu'] = $menu;
                $this->grav['twig']->twig_vars['presentation_breakpoints'] = $breakpoints;
                if ($grav['page']->template() == 'presentation') {
                    $grav['assets']->addInlineCss($this->transport->getStyles(), null, 'presentation');
                }
            }
        }
    }

    /**
     * Handle API
     *
     * @return void
     */
    public function handleAPI()
    {
        $uri = $this->grav['uri'];
        $page = $this->grav['page'];
        $plugins = $this->config->get('plugins');
        if ($uri->path() == '/' . $this->config->get('plugins.presentation.api_route')) {
            if ($_GET['action'] == 'poll') {
                $this->handlePollAPI(
                    $uri,
                    $page,
                    (array) $this->config->get('plugins.presentation')
                );
            }
        }
        if (isset($plugins['admin']) && $plugins['admin']['enabled'] == true) {
            $adminRoute = $this->config->get('plugins')['admin']['route'];
            if ($uri->path() == $adminRoute . '/' . $this->config->get('plugins.presentation.api_route')) {
                if ($_GET['action'] == 'save') {
                    $this->handleSaveAPI();
                }
            }
        }
    }

    /**
     * Handle Save API
     *
     * @return void
     */
    public function handleSaveAPI()
    {
        if (!$this->isAdmin() || empty($_POST)) {
            return;
        }
        header('Content-Type: application/json');
        header("allow-control-access-origin: * ");
        header('HTTP/1.1 200 OK');
        try {
            $post = file_get_contents('php://input') ?? $_POST;
            $post = json_decode($post, true);
            $pages = Grav::instance()['pages'];
            $page = $pages->find('/' . $post['route']);
            $page->rawMarkdown(base64_decode($post['content']));
            $page->save();
            echo '200 OK';
        } catch (\Exception $e) {
            echo $e;
        }
        exit();
    }

    /**
     * Handle Poll API
     *
     * @param array $config Plugin configuration
     *
     * @return void
     */
    public function handlePollAPI($config)
    {
        if (isset($config['sync']) && $config['sync'] == 'poll') {
            set_time_limit(0);
            header('Cache-Control: no-cache, no-store, max-age=0, must-revalidate');
            header('Pragma: no-cache');
            if (!isset($_GET['mode'])) {
                header('HTTP/1.1 400 Bad Request');
                exit('400 Bad Request');
            }
            $res = Grav::instance()['locator'];
            $target = $res->findResource('cache://') . '/presentation';
            include_once __DIR__ . '/API/PollInterface.php';
            include_once __DIR__ . '/API/Poll.php';
            $poll = new Poll($target, 'Poll.json');
            gc_enable();
            if (!isset($config['token']) || empty($config['token'])) {
                return;
            }
            if ($_GET['mode'] == 'set' && isset($_GET['data'])) {
                Utilities::authorize($config['token']);
                $poll->remove();
                header('Content-Type:text/plain');
                header('HTTP/1.1 202 Accepted');
                $poll->set(urldecode($_GET['data']));
            } elseif ($_GET['mode'] == 'get') {
                header('Content-Type: application/json');
                header('HTTP/1.1 200 OK');
                $poll->get();
            } elseif ($_GET['mode'] == 'remove') {
                Utilities::authorize($config['token']);
                header('Content-Type:text/plain');
                header('HTTP/1.1 200 OK');
                $poll->remove();
            }
            $poll = null;
            unset($poll);
            gc_collect_cycles();
            gc_disable();
            exit();
        }
    }

    /**
     * Add templates-directory to Twig paths
     *
     * @return void
     */
    public function templates()
    {
        $this->grav['twig']->twig_paths[] = __DIR__ . '/templates';
    }

    /**
     * Add root URL to Twig vars
     *
     * @return void
     */
    public function twigBaseUrl()
    {
        $uri = $this->grav['uri']->rootUrl(true);
        $this->grav['twig']->twig_vars['presentation_base_url'] = $uri;
    }

    /**
     * Reset cache on shutdown
     *
     * @return void
     */
    public function onShutdown()
    {
        $this->grav['config']->set('system.cache.enabled', $this->cache);
        $this->grav['config']->set('system.pages.never_cache_twig', $this->cacheTwig);
    }

    /**
     * Add Twig Extensions
     *
     * @return void
     */
    public function onTwigExtensions()
    {
        include_once __DIR__ . '/twig/CallStaticExtension.php';
        $this->grav['twig']->twig->addExtension(new PresentationPlugin\CallStaticTwigExtension());
        include_once __DIR__ . '/twig/FileFinderExtension.php';
        $this->grav['twig']->twig->addExtension(new PresentationPlugin\FileFinderTwigExtension());
    }

    /**
     * Register Page templates
     *
     * @param Event $event RocketTheme\Toolbox\Event\Event
     *
     * @return void
     */
    public function onGetPageTemplates(Event $event)
    {
        $types = $event->types;
        $locator = Grav::instance()['locator'];
        $locations = [
            'plugin://' . $this->name . '/blueprints/',
            'theme://blueprints/',
            'user://blueprints/'
        ];
        foreach ($locations as $location) {
            $types->register(
                'presentation',
                $locator->findResource($location . 'presentation.yaml')
            );
            $types->register(
                'slide',
                $locator->findResource($location . 'slide.yaml')
            );
        }
    }

    /**
     * Get API Instance
     *
     * @param string $class   Class name
     * @param mixed  ...$args Class arguments
     *
     * @return mixed Class Instance
     */
    public function getAPIInstance(string $class, ...$args)
    {
        $caller = '\Grav\Plugin\PresentationPlugin\API\\' . $class;
        return new $caller(...$args);
    }

    /**
     * Get list of modular scales
     *
     * @return array List of modular scales
     */
    public static function getModularScale()
    {
        return array(
            ['name' => 'unison', 'ratio' => '1:1', 'numerical' => 1],
            ['name' => 'minor second', 'ratio' => '15:16', 'numerical' => 1.067],
            ['name' => 'major second', 'ratio' => '8:9', 'numerical' => 1.125],
            ['name' => 'minor third', 'ratio' => '5:6', 'numerical' => 1.2],
            ['name' => 'major third', 'ratio' => '4:5', 'numerical' => 1.25],
            ['name' => 'perfect fourth', 'ratio' => '3:4', 'numerical' => 1.333],
            ['name' => 'aug. fourth / dim. fifth', 'ratio' => '1:√2', 'numerical' => 1.414],
            ['name' => 'perfect fifth', 'ratio' => '2:3', 'numerical' => 1.5],
            ['name' => 'minor sixth', 'ratio' => '5:8', 'numerical' => 1.6],
            ['name' => 'golden section', 'ratio' => '1:1.618', 'numerical' => 1.618],
            ['name' => 'major sixth', 'ratio' => '3:5', 'numerical' => 1.667],
            ['name' => 'minor seventh', 'ratio' => '9:16', 'numerical' => 1.778],
            ['name' => 'major seventh', 'ratio' => '8:15', 'numerical' => 1.875],
            ['name' => 'octave', 'ratio' => '1:2', 'numerical' => 2],
            ['name' => 'major tenth', 'ratio' => '2:5', 'numerical' => 2.5],
            ['name' => 'major eleventh', 'ratio' => '3:8', 'numerical' => 2.667],
            ['name' => 'major twelfth', 'ratio' => '1:3', 'numerical' => 3],
            ['name' => 'double octave', 'ratio' => '1:4', 'numerical' => 4]
        );
    }

    /**
     * Parse modular scales for blueprints
     *
     * @return array Blueprint-friendly list of modular scales
     */
    public static function getModularScaleBlueprintOptions()
    {
        $options = ['' => 'None'];
        foreach (self::getModularScale() as $scale) {
            $options[(string) $scale['numerical']] = $scale['numerical'] . ' (' . ucwords($scale['name']) . ', ' . $scale['ratio'] . ')';
        }
        return $options;
    }

    /**
     * Get class names for blueprints
     *
     * @param string $key Needle to search for
     *
     * @return array Blueprint-friendly list of class names
     */
    public static function getClassNamesBlueprintOptions(string $key)
    {
        $regex = '/Grav\\\\Plugin\\\\PresentationPlugin\\\\API\\\\(?<api>.*)/i';
        $classes = preg_grep($regex, get_declared_classes());
        $matches = preg_grep('/' . $key . '/i', $classes);
        $options = [
            '' => 'None',
            $key => $key
        ];
        foreach ($matches as $match) {
            $match = str_replace('Grav\Plugin\PresentationPlugin\API\\', '', $match);
            $options[$match] = $match;
        }
        return $options;
    }

    /**
     * Get reveal.js themes
     *
     * @return array Associative array of styles
     */
    public static function getRevealThemes()
    {
        include __DIR__ . '/vendor/autoload.php';
        $inflector = new Inflector();
        $themes = array('none' => 'None');
        $path = 'user://plugins/presentation/node_modules/reveal.js/css/theme';
        $location = Grav::instance()['locator']->findResource($path, true);
        $files = Utilities::filesFinder($location, ['css']);
        foreach ($files as $file) {
            $key = $file->getBasename('.' . $file->getExtension());
            $themes[$key] = $inflector->titleize($key);
        }
        return $themes;
    }

    /**
     * Initialize shortcodes
     *
     * @return void
     */
    public function onShortcodeHandlers()
    {
        $this->grav['shortcode']->registerAllShortcodes(__DIR__ . '/shortcodes');
    }

    /**
     * Add admin assets
     *
     * @return void
     */
    public function onAdminPagesAssetsInitialized()
    {
        $uri = $this->grav['uri'];
        $res = Grav::instance()['locator'];
        $path = $res->findResource('plugin://' . $this->name, false);
        $adminRoute = $this->config->get('plugins.admin.route');
        if (!Utils::contains($uri->path(), $adminRoute . '/pages')) {
            return;
        }
        if ($this->config->get('plugins.presentation.admin_async_save') !== true) {
            return;
        }
        $adminRoute = $uri->rootUrl(true) . $adminRoute;
        $inlineJsConstants = array(
            'presentationAPIRoute = "' . $adminRoute . '/' . $this->config->get('plugins.presentation.api_route') . '"',
            'presentationAPITimeout = ' . ($this->config->get('plugins.presentation.poll_timeout') ?: 2000) * 2.5,
            'presentationAPIRetryLimit = ' . ($this->config->get('plugins.presentation.poll_retry_limit') ?: 10),
            'presentationAdminAsyncSave = ' . ($this->config->get('plugins.presentation.admin_async_save') ?: 0),
            'presentationAdminAsyncSaveTyping = ' . ($this->config->get('plugins.presentation.admin_async_save_typing') ?: 0)
        );
        $inlineJs = '';
        foreach ($inlineJsConstants as $constant) {
            $inlineJs .= 'const ' . $constant . ';' . "\n";
        }
        $this->grav['assets']->addInlineJs($inlineJs);
        $this->grav['assets']->addJs(
            $path . '/js/save.js'
        );
        $this->grav['assets']->addJs(
            $path . '/node_modules/axios/dist/axios.min.js'
        );
        $this->grav['assets']->addJs(
            $path . '/node_modules/js-base64/base64.min.js'
        );
        $this->grav['assets']->addJs(
            $path . '/node_modules/codemirror/lib/codemirror.js'
        );
    }

    /**
     * Add general assets
     *
     * @return void
     */
    public function onAssetsInitialized()
    {
        if ($this->config->get('plugins.presentation.textsizing')
            && $this->config->get('plugins.presentation.breakpoints')
            && !empty($this->config->get('plugins.presentation.breakpoints'))
        ) {
            $css = '';
            $element = '.reveal .slides section section, .reveal.center .slides section section';
            $breakpoints = array_keys(
                $this->config->get('plugins.presentation.breakpoints')
            );
            $sizes = array_values(
                $this->config->get('plugins.presentation.breakpoints')
            );
            for ($i = 0; $i < count($breakpoints); $i++) {
                $css .= '@media screen and ';
                if ($i == 0) {
                    $css .= '(min-width: 0px) and ';
                    $css .= '(max-width:' . (intval($breakpoints[$i+1])-1) . 'px) ';
                } else {
                    $css .= '(min-width:' . $breakpoints[$i] . 'px) ';
                }
                $css .= '{' . $element . '{font-size:' . $sizes[$i] . 'px !important;}}';
                $css .= "\n";
            }
            $this->grav['assets']->addInlineCss($css, null, 'critical');
        }
        $iframe = '.presentation-iframe {
            width: 100%;
            width: -moz-available;
            width: -webkit-fill-available;
            width: fill-available;
            height: 100%;
            height: -moz-available;
            height: -webkit-fill-available;
            height: fill-available;
          }';
        $this->grav['assets']->addInlineCss($iframe);
    }
}